Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs

```diff
@@ -480,42 +493,161 @@ private PreTypeResolver(ModuleResolver resolver, PreTypeInferenceModuleState pre
+    ///     A ~~ B
+    /// is the disjunction
+    ///     A ::> B    or    B ::> A
+    ///
+    /// If "!forAsOrIs", then "X ::> Y" means
+    ///     X :> Y
+    ///
+    /// If "forAsOrIs", then "X ::> Y" means
+    ///     X' :> Y', or
+    ///     X' and Y' are various bv types, or
+    ///     X' is int and Y' is in {int, char, bv, ORDINAL, real}.
+    /// where X' and Y' are the newtype ancestors of X and Y, respectively.
+    /// Additionally, under the legacy option /generalNewtypes:0 (which will be phased out over time), the latter also allows
+    /// several additional cases, see IsConversionCompatible.
+    /// </summary>
+    bool CheckComparableTypes(PreType a, PreType b, bool forAsOrIs) {
+      if (PreType.Same(a, b)) {
+        // this allows the case where "a" and "b" are proxies that are equal
+        return true;
+      }
+      if (a.Normalize() is not DPreType aa || b.Normalize() is not DPreType bb) {
+        return false;
+      }
+      if (IsSuperPreTypeOf(aa, bb) || IsSuperPreTypeOf(bb, aa)) {
+        return true;
+      }
+      if (!forAsOrIs) {
+        return false;
+      }
+      if (IsConversionCompatible(aa, bb) || IsConversionCompatible(bb, aa)) {
+        return true;
+      }
+      return false;
+    }
+
+    bool IsConversionCompatible(DPreType a, DPreType b) {
+      var aAncestor = AncestorPreType(a);
+      var bAncestor = AncestorPreType(b);
+
+      if (PreType.Same(aAncestor, bAncestor)) {
+        return true;
+      }
+      var aFamily = aAncestor.Decl.Name;
+      var bFamily = bAncestor.Decl.Name;
+      var bName = b.Decl.Name;
+
+      if (IsBitvectorName(aFamily) && (bFamily == "int" || IsBitvectorName(bFamily))) {
+        return true;
+      }
+      if (aFamily == "int" && bName is "char" or "real" or "ORDINAL") {
+        return true;
+      }
+
+      var legacy = !resolver.Options.Get(CommonOptionBag.GeneralNewtypes);
+      if (legacy) {
+        if (aFamily == "real" && (bFamily is "int" or "char" or "ORDINAL" || IsBitvectorName(bFamily))) {
+          return true;
+        }
+        if (aFamily == "char" && (bFamily is "int" or "ORDINAL" || IsBitvectorName(bFamily))) {
+          return true;
+        }
+        if (IsBitvectorName(aFamily) && (bFamily is "int" or "real" or "char" or "ORDINAL")) {
+          return true;
+        }
+      }
+
+      return false;
     }
 
-    bool ApplyComparableConstraints(PreType a, PreType b, IToken tok, string errorFormatString) {
-      // The meaning of a comparable constraint
-      //     A ~~ B
-      // is the disjunction
-      //     A :> B    or    B :> A
+    bool ApproximateComparableConstraints(PreType a, PreType b, IToken tok, bool forAsOrIs, string errorFormatString, bool reportErrors = true) {
+      // See CheckComparableTypes for the meaning of "comparable type".
       // To decide between these two possibilities, enough information must be available about A and/or B.
       var ptA = a.Normalize() as DPreType;
```

Comment:
The prefix `pt` is obsolete in a class already called `PreTypeResolve`. Please rename to `normalizedA` and `normalizedB`